lib/internal/streams/whatwg.js

```diff
@@ -0,0 +1,4233 @@
+    this[kState].reject?.(error);
+  }
+
+  get promise() { return this[kState].promise; }
+}
+
+class ReadableStreamDefaultReader {
+  /**
+   * @param {ReadableStream} stream
+   */
+  constructor(stream) {
+    if (!isReadableStream(stream))
+      throw new ERR_INVALID_ARG_TYPE('stream', 'ReadableStream', stream);
+    this[kState] = {
+      readRequests: [],
+      stream: undefined,
+      close: {
+        promise: undefined,
+        resolve: undefined,
+        reject: undefined,
+      },
+    };
+    setupReadableStreamDefaultReader(this, stream);
+  }
+
+  /**
+   * @returns {Promise<{
+   *   value : any,
+   *   done : boolean
+   * }>}
+   */
+  read() {
+    if (this[kState].stream === undefined) {
+      return PromiseReject(
+        new ERR_INVALID_STATE.TypeError(
+          'The reader is not attached to a stream'));
+    }
+    const readRequest = new DefaultReadRequest();
+    readableStreamDefaultReaderRead(this, readRequest);
+    return readRequest.promise;
+  }
+
+  releaseLock() {
+    if (this[kState].stream === undefined)
+      return;
+    if (this[kState].readRequests.length) {
+      throw new ERR_INVALID_STATE.TypeError(
+        'Cannot release with pending read requests');
+    }
+    readableStreamReaderGenericRelease(this);
+  }
+
+  /**
+   * @readonly
+   * @type {Promise<void>}
+   */
+  get closed() { return this[kState].close.promise; }
+
+  /**
+   * @param {any} reason
+   * @returns {Promise<void>}
+   */
+  cancel(reason) {
+    if (this[kState].stream === undefined) {
+      throw new ERR_INVALID_STATE.TypeError(
+        'The reader is not attached to a stream');
+    }
+    return readableStreamReaderGenericCancel(this, reason);
+  }
+
+  [kInspect](depth, options) {
+    return customInspect(depth, options, 'ReadableStreamDefaultReader', {
+      stream: this[kState].stream,
+      readRequests: this[kState].readRequests.length,
+      close: this[kState].close.promise,
+    });
+  }
+
+  get [SymbolToStringTag]() { return 'ReadableStreamDefaultReader'; }
+}
+
+class ReadableStreamBYOBReader {
+  /**
+   * @param {ReadableStream} stream
+   */
+  constructor(stream) {
+    if (!isReadableStream(stream))
+      throw new ERR_INVALID_ARG_TYPE('stream', 'ReadableStream', stream);
+    this[kState] = {
+      stream: undefined,
+      requestIntoRequests: [],
+      close: {
+        promise: undefined,
+        resolve: undefined,
+        reject: undefined,
+      },
+    };
+    setupReadableStreamBYOBReader(this, stream);
+  }
+
+  /**
+   * @param {ArrayBufferView} view
+   * @returns {Promise<{
+   *   view : ArrayBufferView,
+   *   done : boolean,
+   * }>}
+   */
+  read(view) {
+    if (!isArrayBufferView(view)) {
+      return PromiseReject(
+        new ERR_INVALID_ARG_TYPE(
+          'view',
+          [
+            'Buffer',
+            'TypedArray',
+            'DataView',
+          ],
+          view));
+    }
+    if (view.byteLength === 0 || view.buffer.byteLength === 0) {
```

Comment:
```suggestion
    if (view.byteLength === 0 || ArrayBufferPrototypeByteLength(view.buffer) === 0) {
```
not 100% sure if getters are available in primordials too, but i hope so

also i'm not 100% sure if `view.buffer` is always an ArrayBuffer

Reply:
I do not believe they are. 

If view is an ArrayBufferView then view.buffer should always be an ArrayBuffer.

Reply:
`.buffer` is on DataView.prototype also; it seems ideal to be robust there as well, if possible

Reply:
DataView is acceptable here per the spec. 

It is possible to override `view.buffer` using, for instance, `Object.defineProperty()`, which would cause this to fail. We do not currently cache the getters from these. We could make use of `Relect.get()` here. It adds a bit of complexity tho.